plugins/analysis/common/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/check/Expressions.kt

```diff
@@ -1281,17 +1300,26 @@ private fun SolverState.checkLoopExpression(
 ): ContSeq<StateAfter> =
   when (expression) {
     is ForExpression -> checkForExpression(expression.loopParameter, expression.body, data)
+    is ThreePieceForExpression ->
+      inScope(data) {
+        val initVar = solver.makeObjectVariable(newName(data.context, "initializer", expression))
+        checkBlockExpression(initVar, expression.initializer, false, data).flatMap { after ->
```

Comment:
It would imply quite some refactoring, but it might make things look much prettier. There's some quite nested `flatMap` in the file. But I won't give it much priority, since such a refactoring would need to be quite careful not to omit any `bind`.

Reply:
Right. It’s not possible anyway, forgot for a sec that it’s a multishot data type.